### PR TITLE
 Switch to Scalafmt

### PR DESCRIPTION
**Pull request by patriknw**
* and replace unicode arrows

### CODE REVIEW

**Comment by patriknw** [comment]
One part is missing here, it should also scalafmtOnCompile for multi-jvm sources. Probably some kind of `dependsOn`.

**Comment by patriknw** [comment]
The performance seems ok since it does a good job at format incrementally. 

a full `scalafmt` takes ~50 seconds

If it turns out to be too slow we can disable format on compile, which is what is recommended in docs https://scalameta.org/scalafmt/docs/installation.html#format-on-compile.

Format on save in IntelliJ is probably good to enable, anyway: https://scalameta.org/scalafmt/docs/installation.html#format-on-save

**Review by 2m — APPROVED** [review]
LGTM.

**Comment by 2m** [comment]
There has recently been improvements to sbt-scalafmt's formatting of code in non-standard configurations and also improved cached formatting:

https://github.com/scalameta/sbt-scalafmt/pull/14
https://github.com/scalameta/sbt-scalafmt/pull/13

Improvements not yet released, tho.

**Comment by 2m** [comment]
This should enable the formatting for `MutiJvm` configuration:

```scala
settings(
   inConfig(MultiJvm)(scalafmtConfigSettings)
   ...
)
```

File: .scalafmt.conf
Comment (helena) [inline .scalafmt.conf]: You seem to prefer `AvoidInfix` as well, but we may not want to rewrite those.

Comment (helena) [reply .scalafmt.conf]: My strong pref is for `RedundantBraces` as well but...

File: .scalafmt.conf
Comment (helena) [inline .scalafmt.conf]: `align.tokens = [{code = "=>", owner = "Case"}]` [might be nice](https://scalameta.org/scalafmt/docs/configuration.html#aligntokens)

**Comment by helena** [comment]
Looks good, just a few suggestions to consider, or not.

**Review by helena — APPROVED** [review]
Added just a few things to consider first, otherwise LGTM

File: .scalafmt.conf
Comment (patriknw) [reply .scalafmt.conf]: > RedundantBraces

I think that depends how large the statement is. If it's a real one-liner then yes, but if it's wrapping over several lines and having comments I prefer some extra `{ }` to clarify the scope.

Comment (patriknw) [reply .scalafmt.conf]: I would like to use `AvoidInfix` but it is not nice for some things, like
```
withAttributes(traversalBuilder.attributes.and(attr))
```

but maybe the total result is better? I have pushed that as a separate commit so we can compare and decide: 04ae484

File: .scalafmt.conf
Comment (patriknw) [reply .scalafmt.conf]: thanks, I was looking for that and thought it would be covered by `align = some`.
added in b042e06

**Comment by patriknw** [comment]
> enable the formatting for `MutiJvm` configuration

@2m seems like that is not enough, in the end I'd like it to be done via `test:compile`

File: .scalafmt.conf
Comment (patriknw) [reply .scalafmt.conf]: There were some more trouble with AvoidInfix that I have solved in 32199ce and 1627f14

**Comment by patriknw** [comment]
I will not keep this updated for each and every change in master, so please review although it has conflicts. Before merging I will fix the conflicts ofc.

**Comment by jrudolph** [comment]
> The performance seems ok since it does a good job at format incrementally.
> 
> a full `scalafmt` takes ~50 seconds

What does that mean in practice when switching branches? If the cache is lost in that case, adding 50 seconds to such a build is still a lot.

Disabling format-on-compile somewhat disrupts my workflow but since I'm using IntelliJ anyway, it might still be a net-positive to just integrate it properly into IntelliJ. Since the team mostly uses IntelliJ I guess it would be fine.

Also, on akka-http the automatic pr-formatter [now seems to work](https://github.com/akka/akka-http/pull/2425/commits/e1d35971e341255c401a25edf7f7a1255f0c8ec9), so even if the formatter isn't enabled automatically, we can create such a task for Akka as well and let it auto-format code and amend PRs.

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestKitUtils.scala
Comment (jrudolph) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestKitUtils.scala]: What do we think about automatically adding line breaks like these?

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (jrudolph) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: Why this?

Comment (jrudolph) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: Why this?

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (jrudolph) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: Weird formatting here

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (jrudolph) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: I guess we just accept this style.

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (jrudolph) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: ?

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (jrudolph) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: Weird indenting again. Can we have

```scala
val outcome =
  try fisher(message)
  catch {
  }
```

?

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/javadsl/ManualTime.scala
Comment (jrudolph) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/javadsl/ManualTime.scala]: Good one.

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala
Comment (jrudolph) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala]: Not sure that's how I would break lines here? If it's multiple lines, then rather each argument on an extra line.

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/Effects.scala
Comment (jrudolph) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/Effects.scala]: Can we remove that setting? Seems like a useless change that just might get us more conflicts.

**Review by jrudolph — COMMENTED** [review]
I looked through a few files. A few changes are good but it would be good if we could prevent other ones.

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestKitUtils.scala
Comment (helena) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestKitUtils.scala]: I don't like it here and it ads 4 spaces, should be 2. 
This may be configurable with `verticalMultiline.arityThreshold`

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (helena) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: Let's not allow that format, though it's nice for other things like
```scala
  collection
    .func1(...)
    .func2(...)
    .func3
```

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (helena) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: Let's not do that. Something with `Term`?
https://github.com/scalameta/scalafmt/pull/1187

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (helena) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: I think @patriknw pushes for that over spaces

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (helena) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: +1 that `try catch` line and indent change is bad

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/javadsl/ManualTime.scala
Comment (helena) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/javadsl/ManualTime.scala]: This
```suggestion
  case sc: akka.testkit.ExplicitlyTriggeredScheduler =>
    new ManualTime(sc)
  case _ =>
    throw new IllegalArgumentException(
```

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala
Comment (helena) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala]: Depends on what follows it, e.g.
```scala
 def stop[T](ref: ActorRef[T], max: FiniteDuration = timeout.duration): Unit =
    try Await.result(internalSystem ? { x: ActorRef[ActorTestKitGuardian.Ack.type] =>
```
or

```scala
 def stop[T](ref: ActorRef[T], max: FiniteDuration = timeout.duration): Unit =
    try Await.result(internalSystem ? { 
      x: ActorRef[ActorTestKitGuardian.Ack.type] =>
```

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/Effects.scala
Comment (helena) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/Effects.scala]: +1, the current formatting is nice, it ads that padding

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/CapturedLogEvent.scala
Comment (helena) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/CapturedLogEvent.scala]: 👍

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/CapturedLogEvent.scala
Comment (helena) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/CapturedLogEvent.scala]: better, for 2 args

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/Effect.scala
Comment (helena) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/Effect.scala]: Yes, a space from type def before inner code is good. I see a lot w/out a space.

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/Effect.scala
Comment (helena) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/Effect.scala]: So much better

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/Effect.scala
Comment (helena) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/Effect.scala]: Line after is also so much better

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/StubbedActorContext.scala
Comment (helena) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/StubbedActorContext.scala]: Not a fan of the trailing new line for `)` can we configure that?

**Review by helena — CHANGES_REQUESTED** [review]
Looking good. Added a few  areas to change the configuration on.

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestKitUtils.scala
Comment (patriknw) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestKitUtils.scala]: I would prefer if it stayed on one line if within the 120 chars margin. Can't find a setting for it, do you know?

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (patriknw) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: I agree, I hope it keeps the line breaks if they are there. I'll try
```
optIn.breakChainOnFirstMethodDot = false
```

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (patriknw) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: I prefer `abc.map {` over `abc map {`
For some short operator style method names like `min` it can also be good, but one have to choose.

If we dislike this, we could run one sweep to get rid of infix and then turn off that formatting rule and manually add back infix where it makes sense.

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (patriknw) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: I agree, but how?

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (patriknw) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: maybe `rewrite.neverInfix.excludeFilters` could work

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestKitUtils.scala
Comment (patriknw) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestKitUtils.scala]: I think it's actually doing the right thing here, the line exceeds 120 and then that is a good place to break it.

I was worried that it always breaks before `extends`

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (patriknw) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: it's 119 chars, so fits on one line

(I would have expected it to preserve existing line breaks though)

File: .scalafmt.conf
Comment (helena) [reply .scalafmt.conf]: `AvoidInfix` seems good now, WDYT?

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (patriknw) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: now it's not touching this, not sure what I changed, perhaps the version

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (patriknw) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: fixed this manually before formatting

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/Effects.scala
Comment (patriknw) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/Effects.scala]: fixed: `spaces.inImportCurlyBraces = true`

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/StubbedActorContext.scala
Comment (patriknw) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/StubbedActorContext.scala]: I agree, fixed: `danglingParentheses = false`

File: .scalafmt.conf
Comment (patriknw) [reply .scalafmt.conf]: `rewrite.neverInfix.excludeFilters` worked great, now it's `AvoidInfix` with some listed exceptions

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (patriknw) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: didn't work can't find the setting for that, we would have to ask

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (patriknw) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: couldn't find a cure for this one

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (patriknw) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: `rewrite.neverInfix.excludeFilters` worked great

**Comment by patriknw** [comment]
adjusted config some more: 355649e 
and updated to newer version: 49357b6

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestKitUtils.scala
Comment (dwijnand) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestKitUtils.scala]: > I don't like it here and it ads 4 spaces, should be 2. Perhaps the default is 4.
> `continuationIndent.defnSite = 2`
> ?

Bodies are indented 2, so continuations being indented 4 is a nice visual guide that it's not the body.  IMHO

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (dwijnand) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: Yeah, long-standing bug.

File: .scalafmt.conf
Comment (dwijnand) [reply .scalafmt.conf]: "some" :D

**Comment by patriknw** [comment]
@2m my bad, your advice for formatting for `MutiJvm` worked, thanks

**Comment by raboof** [comment]
> a full `scalafmt` takes ~50 seconds
> 
> If it turns out to be too slow we can disable format on compile, which is what is recommended in docs https://scalameta.org/scalafmt/docs/installation.html#format-on-compile.

It does look like switching to another branch makes the formatting run again... not sure if we want to accept that. OTOH perhaps it is OK to disable format-on-compile since scalafmt editor plugins are more ubiquitous?

I tried with sbt-scalafmt master and https://github.com/scalameta/sbt-scalafmt/pull/13 indeed seems to make a large difference. Filed https://github.com/scalameta/sbt-scalafmt/issues/19.

**Comment by patriknw** [comment]
@raboof Let's use it without format on compile to avoid slowdown and that's also what is recommended by scalafmt docs. Format on save from IntelliJ instead.

I have also added formatting to the all* command aliases.

Defined version to 2.0.0-RC5

and rebased

**Ready for final approval or rejection**

BTW, I will squash the commits before merging, but keeping the "format source" separate and also see if I can use another bot author for that commit (I don't want to take credit for that big changes when it's only by tool).

**Comment by helena** [comment]
@patriknw what about anyone not using IntelliJ - if you're making the assumption.

**Comment by dwijnand** [comment]
Setting the version to 2.0.0-RC5 and then using sbt-scalafmt 1.6.0-RC4 is probably not great.  Note that there is no version of sbt-scalafmt for Scalafmt 2 yet (tracked as https://github.com/scalameta/sbt-scalafmt/issues/15).

For Play, I opted for using Scalafmt 2 by configuring CI to use the CLI launcher.

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala
Comment (jrudolph) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala]: Agreed

File: build.sbt
Comment (jrudolph) [inline build.sbt]: That's nice!

**Comment by jrudolph** [comment]
@patriknw can you sum up how the new workflow will be? (It's hard to say because build formatting changes makes it hard to review any actual build changes).

 * How does the developer use it? Usually, with IntelliJ IIUC?
 * How do we enforce correct formatting in PRs? Does it fail fast?
 * Do we need to update any contribution docs, so that contributors will by default do the right thing?

**Comment by patriknw** [comment]
@dwijnand Thanks for clarifying that. I thought it was supported because the tickets mentioned that a release was already done. I was some other tickets mentioning scalafmt-dynamic and guessed it was using that already since I couldn't find any reference to a scalafmt version anywhere.

> probably not great

What does that mean? It doesn't fail so is it just that it's not using the specified version?

What version is supported by the IntelliJ plugin?

**Comment by dwijnand** [comment]
> What does that mean? It doesn't fail so is it just that it's not using the specified version?
> 
> What version is supported by the IntelliJ plugin?

Only as of Scalafmt 2.0.0-RC5 does `scalafmt-dynamic` use the version defined in `.scalafmt.conf` to dispatch to the right version of Scalafmt.  https://github.com/scalameta/scalafmt/commit/fd48c4b0ea33608923dcbf78ec4bfafdb06699bd

IntelliJ currently supports latest 1.5 (1.5.1) and latest 1.6 RC (1.6.0-RC4).  I think development on the 1.x line has stopped and the setup around the 2.x line is coming.  For instance, https://github.com/scalameta/scalafmt/commit/fd840e7d8e4d1371e731447e3baf710a0343a7b5 where work was done for the IntelliJ integration.

So you can either:
* hold this PR, wait for 2.x to be fully ready, then merge
* merge this PR, as 2.x becomes more ready move to a preferred workflow
* switch to 1.6, reformat, then when 2.x is ready upgrade (and probably reformat some differently-formatting sources)

**Comment by 2m** [comment]
The sbt plugin with enhancements has been released. Note that org for the artifacts has changed as well:

```scala
addSbtPlugin("org.scalameta" % "sbt-scalafmt" % "2.0.0-RC5")
```

**Comment by dwijnand** [comment]
Yeah, I just resolved it locally.  I'm switch play-grpc to it.  The check tasks are `scalafmtCheckAll` and `scalafmtSbtCheck`.

**Comment by dwijnand** [comment]
The IntelliJ plugin has also been updated: https://plugins.jetbrains.com/plugin/8236-scalafmt/update/59101

So the concerns I had shared here to forewarn about adopting Scalafmt 2 have all been resolved.  🕺

**Comment by patriknw** [comment]
I have updated to sbt-scalafmt 2.0.0-RC5 and rebased.
I tried the IntelliJ nigtly plugin and it seems to work.

@jrudolph The workflow is to use the IntelliJ plugin and configure it to format on save. Then nothing else should be needed. I added note in CONTRIBUTING about the plugin.

If not using the plugin the sources can be formatted from `sbt` with `scalafmtAll`, typically on individual projects that you have changed. I added note in CONTRIBUTING about this.

I also added `scalafmtAll` to the `all*` command aliases.

We currently have a `git diff` check in jenkins pr validator job. That can be replaced by `clean scalafmtCheckAll test:compile`, which will fail the build earlier than before. We probably have to keep the `git diff` to detect java formatting issues.

**Review by johanandren — APPROVED** [review]

**Comment by patriknw** [comment]
RoutersSpec

**Comment by patriknw** [comment]
PLS BUILD

**Comment by johanandren** [comment]
@patriknw that RoutersSpec has been failing all weekend I have fixed it in a PR today so maybe just go ahead and merge anyways?

**Comment by patriknw** [comment]
I will not merge this (or anything) without green PR validation.

**Comment by johanandren** [comment]
In that case, could be good to merge #26497 which fixes the `RoutersSpec`

**Pull request merged**
